### PR TITLE
Add recommended payment methods

### LINKED ISSUE #6684: Payments settings: add recommended payment methods
**Description:** we have a significant % of new users going to the payments settings without visiting the payments task or before installing a payments extension. This is a good moment to surface important payment options to medium/high intent users.

Settings - WC Pay

**Acceptance criteria:** 
- [ ] Introduce a new card in the Payments settings with the WCPayments, Stripe and PayPal as recommendations
- [ ] The card is displayed regardless they have a payments extension installed
- [ ] Ideally we should use the system that exists for marketplace suggestions. I'm assuming that this will allow us to tweak the content without relying on a core release
- [ ] Include a "See more options" link at the bottom of the card. The link should redirect users to https://woocommerce.com/product-category/woocommerce-extensions/payment-gateways/?utm_source=payments_recommendations
- [ ] Automatically install the extension if users click "Get started". I'm not sure if this is possible if feed the card through the marketplace
- [ ] Display it only in countries where WCPayments is available: US, AU, CA, GB, IE and NZ(let's assess the effectiveness of this approach before having it in all countries)
- [ ] The user can dismiss the recommendations card through the ellipsis icon
- [ ] Don't show the suggestions card if the user has the marketplace suggestions settings disabled (settings > advanced > woocommerce.com > Marketplace suggestions > Show Suggestions)

**Event tracking:**
`wcadmin_settings_payments_recommendations_setup`
When the user clicks "get started"
Event prop: `extension_selected`: extension name

`wcadmin_settings_payments_recommendations_dismiss`
When the user clicks "dismiss"

`wcadmin_settings_payments_recommendations_pageview`
When the user visits the payments settings with the recommendations card enabled

### PR DESCRIPTION
**Pull request by louwie17**
Fixes #6684 

This PR depends upon 9916-gh-Automattic/woocommerce.com  being merged.
The main change for this PR is adding a new card under the payments settings list with recommended payment providers, the list is retrieved from a woocommerce.com json, added in the PR mentioned above.

Additional changes:
- Add an embedded layout component that gets appended below the `wpBody` component, I added support for what I call a registry, allowing us to register react components for a specific page, see the [payment recommendations component](https://github.com/woocommerce/woocommerce-admin/blob/6f0e3c457469857b519b16d91f8f3476f69b1018/client/payments/index.ts#L18-L21) as an example. I added a custom history item that appends the `page`, `tab`, and/or `section` together with an `_`, allowing us to target the specific pages more easily by using the react router.
- Added a `useOptionsHydration` hook, which is mostly a wrapper of the `withOptionsHydration` function.
- Converted some (probably should finish) of the plugins data package to TS, and added support for the recommended payments query.

### Accessibility

-   [x] I've tested using only a keyboard (no mouse)
-   [ ] I've tested using a screen reader
-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

Recorded on my local dev, it loads quite a bit faster on jurassic:
recommended-payments

### Detailed test instructions:
  
- If 9916-gh-Automattic/woocommerce.com is not merged yet (Merged now, so don't need to update this), replace the `$url` [here](https://github.com/woocommerce/woocommerce-admin/blob/6f0e3c457469857b519b16d91f8f3476f69b1018/src/PaymentPlugins.php#L50) with this: `https://gist.githubusercontent.com/louwie17/f5c0a1c57d3a9ead78fcf385782d8a05/raw/20809ef91680f86d73f47caf6349055512ec695f/recommended-payments.json`, note that the icons won't show unless you also serve up the WooCommerce.com PR mentioned earlier. And start your local environment or generate a `test:zip` for jurassic
- Make sure your store location is filled out and within US | PR | AU | CA | GB | IE | NZ and enable debug messages in console using `localStorage.setItem( 'debug', 'wc-admin:*' );`
- Visit **Woocommerce > Settings > Payments** you might have to wait a couple seconds, but it should show a card with **Recommended ways to get paid** listing 3 different payment providers (this data comes from the json referred to earlier).
- It should display a `wcadmin_settings_payments_recommendations_pageview` track
- Click `Get started` on one of the providers, it will show a loading icon (installing the plugin), once done it should redirect you to the plugin set up page.
- It should trigger a `wcadmin_settings_payments_recommendations_setup` track with an `extension_selected` attribute set to the extension slug.
- Go back to the payment settings page
- Notice how the plugin you had previously installed and activated does not display anymore.
- Click on the 3 dots of the card, click `Hide this`, it should make the card disappear, and also not be displayed on refresh. Deleting `woocommerce_show_marketplace_suggestions` option, should make it appear again.
Hiding the card should trigger a `wcadmin_settings_payments_recommendations_dismiss` track.

After making sure `woocommerce_show_marketplace_suggestions` is deleted and the card shows up again.
- Go to **WooCommerce > Settings > Advanced > WooCommerce.com** and un-select **Show Suggestions** and save
- Go to the payments setting screen again, the card should not be displayed, nor should any requests have gone out to recommended-payments (it should get the component chunk though).

Enable the **Show Suggestions** in the **Woocommerce.com** settings again, and change the store's country in **WooCommerce > Settings > General** to something not in this list: US | PR | AU | CA | GB | IE | NZ. (You can pick France). 
- Go back to the Payment settings screen, the recommended card should not show up.

### CODE REVIEW

File: .eslintrc.js
Comment (louwie17) [inline .eslintrc.js:39]: As mentioned in the comment, this fixes a weird error when exporting enums.

File: client/embedded-body-layout/embedded-body-layout.tsx
Comment (louwie17) [inline client/embedded-body-layout/embedded-body-layout.tsx:52]: Overwriting the location param to create a path out of the query params `page`, `tab`, and `section`.

File: client/embedded-body-layout/page-registry.ts
Comment (louwie17) [inline client/embedded-body-layout/page-registry.ts:21]: See use of this in `client/payments/index.ts`

File: packages/data/src/options/with-options-hydration.js
Comment (louwie17) [inline packages/data/src/options/with-options-hydration.js:13]: Added a hook for it as well, which I was going to use, but then didn't need to in the end.

File: packages/data/src/types.ts
Comment (louwie17) [inline packages/data/src/types.ts:21]: See comment, we can use this for exporting selector types, this omits the first parameter (state), which is not necessary when calling the selector within a component.

File: src/API/Plugins.php
Comment (louwie17) [inline src/API/Plugins.php:466]: Not used at the moment, but if we decide to support multiple languages, we can easily add it to the `json` now.

File: typings/global.d.ts
Comment (louwie17) [inline typings/global.d.ts:9]: Copied this comment from TS docs

File: packages/data/src/types.ts
Comment (samueljseay) [reply packages/data/src/types.ts:21]: Yep great call on this one 👏

File: packages/data/src/plugins/action-types.ts
Comment (samueljseay) [inline packages/data/src/plugins/action-types.ts:12]: Can we avoid default exports? naming them avoids confusion and has better IDE support.

File: src/API/Plugins.php
Comment (samueljseay) [inline src/API/Plugins.php:449]: should be "recommended"?

File: client/payments/payment-recommendations.tsx
Comment (samueljseay) [inline client/payments/payment-recommendations.tsx:89]: There is a lot of logic here. This really needs a test if one isn't there already. I did a quick text search and didn't see one.

File: client/payments/payment-recommendations.tsx
Comment (samueljseay) [inline client/payments/payment-recommendations.tsx:106]: This looks good. Not a required change, but seems like it'd be a great hook to introduce, because we do a lot of this kind of analytics.

File: client/payments/payment-recommendations.tsx
Comment (samueljseay) [inline client/payments/payment-recommendations.tsx:92]: Great job on this component. its looking really good. There's a lot going on but it does a lot too. Nice one!

**Review by samueljseay — COMMENTED** [review]
This is overall looking great. A lot of great additions to the typing information as well. Just a few (mostly small) suggestions from me.

The biggest suggestion I'd be keen to address is if we could use SlotFill instead of a custom page registry.

File: src/API/Plugins.php
Comment (louwie17) [reply src/API/Plugins.php:449]: It's not the first time I spelled it like that haha

File: client/payments/payment-recommendations.tsx
Comment (louwie17) [reply client/payments/payment-recommendations.tsx:89]: There should be tests for this, maybe I forgot to push it up. This is why I end up exporting this function.

File: packages/data/src/plugins/test/reducer.ts
Comment (louwie17) [reply packages/data/src/plugins/test/reducer.ts:24]: Do you have any suggestions here? as the issue is the `test` or I should remove the de-structuring in the reducer. 
I could make all the items in `Actions` optional, but then I will have to handle all the optional cases within the switch statement, which doesn't quite make sense.
The ideal would be, that TS could determine the action key/values by the `type` value, but I can't find a good way to do that without writing specific types for each action (I was hoping to avoid that and be able to use the ReturnType).

So I be happy with an alternate solution?

File: packages/data/src/plugins/test/actions.ts
Comment (louwie17) [reply packages/data/src/plugins/test/actions.ts:120]: Lint doesn't like that, I will just replace with `jest.fn()` instead.

File: client/payments/payment-recommendations.tsx
Comment (louwie17) [reply client/payments/payment-recommendations.tsx:89]: Tested here: https://github.com/woocommerce/woocommerce-admin/blob/feature/6684_add_recommended_payment_methods/client/payments/test/payment-recommendations.test.tsx#L90-L245

**Comment by louwie17** [comment]
> This is overall looking great. A lot of great additions to the typing information as well. Just a few (mostly small) suggestions from me.

I believe I covered most of your suggestions, although curious if you have some alternate suggestions for: https://github.com/woocommerce/woocommerce-admin/pull/6760#discussion_r611248963

> The biggest suggestion I'd be keen to address is if we could use SlotFill instead of a custom page registry.

I don't think `SlotFill` would be very beneficial here, given we still render the component within the same element tree, everything as a direct child of the embedded layout component.
I could probably remove the page registry and just render the payments component within the `Switch` component, if you prefer that? Given I have to `import '../payments'` there anyway.

File: packages/data/src/plugins/test/reducer.ts
Comment (samueljseay) [reply packages/data/src/plugins/test/reducer.ts:24]: @louwie17 yep for sure. If `{}` is a legal argument then you simply union it with Action in the signature of the reducer, then you can use the switch statement to do type narrowing and you don't need to declare anything as optional.

Code example:

```js
type MyCoolAction = {
    type: 'cool-action';
    baz: string
};

type MyOtherCoolAction = {
    type: 'my-other-one';
    foo: number
}

type Action = MyCoolAction | MyOtherCoolAction


const reducer = (action: Action | {}) => {
    if ("type" in action) {
        switch (action.type) {
            case 'cool-action':
                console.log(action.baz);
                return {};
            case 'my-other-one': {
                // error
                console.log(action.baz);
                // ok
                console.log(action.foo);
                return {};
            }   
        }
    } else {
        return {};
    }
}
```

Here's the TS playground link if you want to have a play with it: https://www.typescriptlang.org/play?ssl=31&ssc=2&pln=1&pc=1#code/C4TwDgpgBAsiDCB7RAbAggY2AS0QOygF4oBvAKCkqlEgC4oByDZFAWgEMtc8GBuCqgCN2AL3oBnYACdseAOZkAvvzI1ocAPLAAFhClJUmHPiKkBlNfQYBbEK0Q699vBD7moAM2T08AV2uCekpkquDQRtymcAboXCYAPrAgWrr6LBH4IWTMeJJQUhAAJr4YeqYAFJzGePQZBIkkigCURAB8ZlRQ2B5Q5QBEan1dBFXcLeSdneIA7tjAGNq9o-gAdGrj7pNUGOzi0EwsHHE8tJtbkzniqBArKIhylccrwiJN-Ocf+RDAvlIEje9PpQdntGLZ7I4pM5XPQJkDJgB6BFQPRSRBSM5Ay7XW73R7VZ6iN6Yz5IqCIADWJI+2JQNzuD2WeBWXkQxPhkwKPz+pGU1KoiioJMU7kFEBQoLhWy5v3+fM6IpFZCAA

File: packages/data/src/plugins/test/actions.ts
Comment (samueljseay) [reply packages/data/src/plugins/test/actions.ts]: That seems like an annoying rule 🤔 I wonder if we could override that rule for the test directory. Something we can look at in future.

**Comment by samueljseay** [comment]
> I don't think SlotFill would be very beneficial here, given we still render the component within the same element tree, everything as a direct child of the embedded layout component.
I could probably remove the page registry and just render the payments component within the Switch component, if you prefer that? Given I have to import '../payments' there anyway.

I'm not gonna die on this hill, I trust you will make the pragmatic decision but I'll just add some additional thoughts and leave it up to you:

1. I was thinking that although we're not rendering outside the tree one benefit is, its an existing API thats simpler and we're already using it so no extra code to maintain.

2. It may be desirable to let other developers like internal plugin teams extend it in future, in which case SlotFill is way way better.

I respect your reasoning though! Your call ❤️

File: packages/data/src/plugins/test/reducer.ts
Comment (louwie17) [reply packages/data/src/plugins/test/reducer.ts:24]: Thanks @samueljseay :) that did the trick (almost) haha I did have to make sure I add a return type to each action specifying the specific action enum it used. Otherwise type stayed a generic `ACTION_TYPES` and the attributes wouldn't be passed down into the switch statement.
Much nicer this way though.

Comment (louwie17) [reply packages/data/src/plugins/test/reducer.ts]: Fixed here: https://github.com/woocommerce/woocommerce-admin/pull/6760/commits/7bddb6428d44ba15d343e9b52b269ecb36aa2b6e

**Comment by louwie17** [comment]
> I was thinking that although we're not rendering outside the tree one benefit is, its an existing API thats simpler and we're already using it so no extra code to maintain.

I would argue that my API is pretty simple, but 'yes' an extra item to maintain.

> It may be desirable to let other developers like internal plugin teams extend it in future, in which case SlotFill is way way better.

This is a valid point, which makes SlotFill a valid point indeed, I will play around with it and let you know how I make out.

**Comment by louwie17** [comment]
@samueljseay I did add SlotFill in this commit: https://github.com/woocommerce/woocommerce-admin/pull/6760/commits/5fc3e2ed40193d6cf852fddff9ae0dadb39e38cf
It seems fine, I actually like pushing the show/hide logic to the Fill component instead of using the router. 

I am fine with sticking with this for now, until the use case expands.

File: packages/data/package.json
Comment (louwie17) [inline packages/data/package.json:26]: The `woocommerce/date` package relies on an older version of `i18n`, which has a wrong TS type, the newer version has the correct type.

File: readme.txt
Comment (octaedro) [inline readme.txt:92]: Hi @louwie17,
Just chiming because I think that there is a typo here: `stetings`.

File: client/payments/payment-recommendations.tsx
Comment (jeffstieler) [inline client/payments/payment-recommendations.tsx:124]: Since we're only allowing one plugin install at a time, should we disable the other buttons while installing?

File: client/payments/payment-recommendations.tsx
Comment (jeffstieler) [inline client/payments/payment-recommendations.tsx:119]: I _think_ we've standardized on using the strings `yes` and `no` for boolean options.

File: client/payments/payment-recommendations.tsx
Comment (jeffstieler) [inline client/payments/payment-recommendations.tsx:150]: Missing translation.

File: src/API/Plugins.php
Comment (jeffstieler) [inline src/API/Plugins.php:454]: The worst nitpick ever - "plugin" isn't pluralized. I know, I'm terrible.

File: src/API/Plugins.php
Comment (jeffstieler) [inline src/API/Plugins.php:459]: I'm not 100% sure locale is the correct value to use for comparison here (IIRC that can differ from the store's location). AFAIK we've been using `wc_get_base_location()` elsewhere for checking what country a store is in.

File: src/PaymentPlugins.php
Comment (jeffstieler) [inline src/PaymentPlugins.php:54]: This should be `wp_safe_remote_get()`.

File: src/PaymentPlugins.php
Comment (jeffstieler) [inline src/PaymentPlugins.php:80]: Adding an `updated` timestamp could be helpful for debugging. I'm lifting this idea straight from https://github.com/woocommerce/woocommerce/blob/master/includes/admin/marketplace-suggestions/class-wc-marketplace-updater.php which is essentially the same type of operation.

File: src/API/Plugins.php
Comment (jeffstieler) [inline src/API/Plugins.php:467]: Not saying we need to utilize this, but I was searching for an existing function that might hold this business logic: [`WC_Marketplace_Suggestions::allow_suggestions()`](https://github.com/woocommerce/woocommerce/blob/master/includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php#L161)

**Review by jeffstieler — CHANGES_REQUESTED** [review]
I should point out right away that I didn't really review the TypeScript changes since I'm out of my depth and @samueljseay already did.

Other than the feedback left inline, I think we might be in violation of the WP.org guidelines since we do not explicitly call out that clicking "Get started" will install a plugin. cc: @pmcpinto @elizaan36

File: src/API/Plugins.php
Comment (louwie17) [reply src/API/Plugins.php:467]: Yeah that's where I took it from. My logic is currently a little different, to allow for new languages more easily, but I suppose I could leave that out.

File: client/embedded-body-layout/embedded-body-layout.tsx
Comment (louwie17) [reply client/embedded-body-layout/embedded-body-layout.tsx:27]: Yeah I had mixed feelings about this, originally I was using a filter combined with Router. The Router part isn't necessary, and I might just replace the Slot Fill with a filter as I think that is a better option.

File: src/API/Plugins.php
Comment (louwie17) [reply src/API/Plugins.php:459]: This is referring to the language though, as the json data is in english, which is why we only support english locales.

Comment (jeffstieler) [reply src/API/Plugins.php:459]: Ah gotcha, I didn't catch that.

File: client/embedded-body-layout/embedded-body-layout.tsx
Comment (louwie17) [reply client/embedded-body-layout/embedded-body-layout.tsx]: Removed this now, in favour of using `applyFilters` done in this commit -> https://github.com/woocommerce/woocommerce-admin/pull/6760/commits/41ea96749d296a17e422ac2dfbc680f35ad4b39e

File: client/payments/index.ts
Comment (jeffstieler) [inline client/payments/index.ts:1]: No more slot. 😄

**Comment by elizaan36** [comment]
> Other than the feedback left inline, I think we might be in violation of the WP.org guidelines since we do not explicitly call out that clicking "Get started" will install a plugin. cc: @pmcpinto @elizaan36

@jeffstieler We could easily update the copy but this is the same pattern used in the "Choose payment methods" task screen for a while now (maybe ever since the task list was implemented?) Choosing "Get started" installs the extension with no warning and launches into the setup flow.

@louwie17 one comment on the "See more options" action in the card footer. Can we instead make that a tertiary button (link only). Want to visually differentiate that from the others by de-prioritizing it a bit. 

Small thing as well but it would be best to left-align the card action to be consistent with the primary action of the screen. (And to make sure it shows up on mobile.)

**Comment by louwie17** [comment]
> @elizaan36 @louwie17 considering that we're displaying the recommended card on a permanent screen (settings) and that we're pulling the content from a remote json file, I think it's better to be extra cautious and include the following copy: The extension will be installed and activated for you when you click "Get started". Sounds good?

The copy sounds good to me, where would you want this added? added on to each plugin's copy, or more something like the business feature page, as copy below the card?
Screen Shot 2021-04-20 at 9 17 56 AM

File: client/payments/payment-recommendations.tsx
Comment (louwie17) [reply client/payments/payment-recommendations.tsx:124]: Yeah we probably should

**Comment by elizaan36** [comment]
I think placing the copy in the card header makes sense as you suggested @louwie17. We can add it to the description copy rather than duplicating for each extension. Also I think below the card will be too hidden.

Here's one way of adding the copy:
image

**Comment by louwie17** [comment]
Thanks @elizaan36 & @pmcpinto , here is the updated screenshot:
Screen Shot 2021-04-20 at 9 55 11 AM

File: client/payments/payment-recommendations.scss
Comment (jeffstieler) [inline client/payments/payment-recommendations.scss]: Nitpick - the "get started" buttons could use some margin/padding to the left.

**Review by jeffstieler — APPROVED** [review]
Pre-approving - some minor tweaks noted. 👍

**Comment by louwie17** [comment]
> Looks good, just curious if we can reduce the width of the text fields for each extension. Those lines of text are very long.

Yeah I can.

> @louwie17 would you mind sharing a screenshot of the mobile view as well?

Sure here it is (from my iphone):
IMG_B7329B85D2DC-1

**Comment by elizaan36** [comment]
So perfect, thanks!

Just needs the period after "Get started" and it's gtg🚢

**Comment by mattsherman** [comment]
> IMG_B7329B85D2DC-1

Some little drive-by design thoughts that came into my mind when I saw this in my inbox...

- It seems a little odd to have the header say "**Recommended** ways of getting paid" and then have a single item also say "**Recommended**"
- The "Recommended" pill looks kinda cluttered in the mobile view... have you considered putting it above the payment method name ("WooCommerce Payments")?

**Comment by elizaan36** [comment]
thanks for the design comments, Matt.

> It seems a little odd to have the header say "Recommended ways of getting paid" and then have a single item also say "Recommended"

The goal is to differentiate WCPay in the list and this is component is a subtle way of doing so. There is some redundancy to say recommended twice, but I don't think it's a huge problem. We can explore other labels such as "Built by WooCommerce" in the future if this label seems to be a problem with adoption.

> The "Recommended" pill looks kinda cluttered in the mobile view... have you considered putting it above the payment method name ("WooCommerce Payments")?

There could be some top and bottom padding on the pill in the mobile view, but I wouldn't place the recommended pill above the extension title. Having the titles in the same place/orientation creates visual consistency that makes it easier to scan the page.

**Pull request merged**
